working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,285 @@
+Version: 0.3 (draft)
+
+## Background
+
+See issue #698 for the feature discussion.
+
+### Verbosity
+
+Creating a class with a simple constructor is fairly short because of Dart's initializing formals, but it's still entirely boiler-plate code that repeats names that already exist elsewhere in the class declaration. You need to repeat the name at least once. Forwarding constructor parameters is even more verbose, especially for named parameters.
+
+```dart
+class Complex {
+  final double real, imaginary;
+  Complex({this.real, this.imaginary});
+}
+class ColorComplex extends Complex {
+  final Color color;
+  ColorComplex({double real, double imaginary, this.color})
+      : super(real: real, imaginary: imaginary);
+      // "real"/"imaginary" written three times.
+}
+```
+
+It would be nice to have a simpler way to handle simple cases, without making anything harder for complex cases.
+
+### Default Constructors
+
+Dart adds a *default constructor* to any class which doesn't declare any constructors. The default constructor is an unnamed generative constructor which is equivalent to `ClassName() : super();`.
+
+If that constructor would not be valid for the class, for example if the class declares a final instance variable with no initializer, or if the superclass does not have an unnamed generative constructor with no required parameters, then it's a compile-time error.
+
+In most situations, there is no difference between declaring the constructor above explicitly and getting the default constructor. The only distinction is that you can extract a mixin from a class if it has `Object` as superclass and it *declares* no non-factory constructor.
+
+### Mixin Application Forwarding Constructors
+
+A *mixin application* receives corresponding forwarding constructors for all accessible generative constructors of the superclass. The forwarding constructor has the same basename (identifier after the class name for named constructors, empty for the unnamed constructor), the same parameter signature as the superclass constructor, including default values for optional parameters, and an initializer list with a `super`-invocation to the corresponding superclass constructor which forwards each parameter as an argument to the corresponding superclass constructor parameter. The forwarding constructor is `const` if the superclass constructor is `const` *and* the mixin declares no instance variables.
+
+Since a mixin application class cannot contain static members, there is no risk of the introducing a conflict between a constructor name and a static member. 
+
+### Goal
+
+We want to allow classes with fields without having to write constructors. That is, we want *initializing default constructors* which takes parameters for class fields (with some necessary exceptions) and initializes the fields.
+
+We want those constructors to be independent of the order of the fields in the class, so the constructor parameters will be *named*.
```

Comment:
Does this mean that private fields are not allowed? We should at least call that out if so.

If they are allowed, that introduces cases where adding a _private_ field in a class is now a breaking change - because the name would collide with a public or private field initializer in any class that extends that class.